eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -34,75 +34,45 @@ import fr.acinq.eclair.{wire, _}
 import kamon.Kamon
 import scodec.bits.ByteVector
 
-import scala.concurrent.duration._
-import scala.util.Random
-
 /**
  * This actor represents a logical peer. There is one [[Peer]] per unique remote node id at all time.
  *
- * The [[Peer] actor is mostly in charge of managing channels, stays alive when the peer is disconnected, and only dies
+ * The [[Peer]] actor is mostly in charge of managing channels, stays alive when the peer is disconnected, and only dies
  * when there are no more channels.
  *
  * Everytime a new connection is established, it is sent to the [[Peer]] and replaces the previous one.
  *
  * Created by PM on 26/08/2016.
  */
-class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, router: ActorRef, watcher: ActorRef, relayer: ActorRef, paymentHandler: ActorRef, wallet: EclairWallet) extends FSMDiagnosticActorLogging[Peer.State, Peer.Data] {
+class Peer(val nodeParams: NodeParams, remoteNodeId: PublicKey, switchboard: ActorRef, router: ActorRef, watcher: ActorRef, relayer: ActorRef, paymentHandler: ActorRef, wallet: EclairWallet) extends FSMDiagnosticActorLogging[Peer.State, Peer.Data] {
 
   import Peer._
 
+  val reconnectionTask: ActorRef = context.actorOf(ReconnectionTask.props(nodeParams, remoteNodeId, switchboard, router), "reconnection-task")
+  // we don't use FSM.SubscribeTransitionCallBack(reconnectionTask) because we don't the reconnectionTask to miss the first transition
```

Comment:
Can you clarify? It's the initial going to `INSTANTIATING` that we don't want to miss?

nit: I think this sentence is missing "want":

```suggestion
  // we don't use FSM.SubscribeTransitionCallBack(reconnectionTask) because we don't want the reconnectionTask to miss the first transition
```

Reply:
Here is what happens when you send an `FSM.SubscribeTransitionCallBack` to an FSM:

```scala
    case SubscribeTransitionCallBack(actorRef) ⇒
      // TODO Use context.watch(actor) and receive Terminated(actor) to clean up list
      listeners.add(actorRef)
      // send current state back as reference point
      actorRef ! CurrentState(self, currentState.stateName)
```

If we make the `ReconnectionTask` subscribe, or even have the `Peer` subscribe its child actor by sending a message to itself, there would be a race between `FSM.SubscribeTransitionCallBack` and `Peer.Init` (and potentially other messages like a `Peer.Connect`). This would complicate the `ReconnectionTask` because the first transition would be unpredictable. Sure, we would be able to sort things out by looking at the `CurrentState` but it's just more work.

Making the `Peer` synchronously register the `ReconnectionTask` as a listener removes the race condition. The `ReconnectionTask` can now safely assume that it will always receive the same initial transitions (`INSTANTIATING`->`INSTANTIATING` and then `INSTANTIATING`->`DISCONNECTED`). We can then get rid of the `CurrentState` handling.

Reply:
Ok, that makes sense. Maybe we can make the comment clearer, something like: `we register the reconnection task to our transitions ourselves; if we let the child actor register itself, there is a race condition and it may miss the first transition.`?